File: appplnr/PnlPlnrSruRec.cpp
/**
  * \file PnlPlnrSruRec.cpp
  * app access code for job PnlPlnrSruRec (implementation)
  * \author Alexander Wirthmueller
  * \date created: 4 Dec 2015
  * \date modified: 4 Dec 2015
  */

#include "PnlPlnrSruRec.h"

/******************************************************************************
 class PnlPlnrSruRec::VecVDo
 ******************************************************************************/

uint PnlPlnrSruRec::VecVDo::getIx(
			const string& sref
		) {
	string s = StrMod::lc(sref);

	if (s.compare("butminimizeclick") == 0) return BUTMINIMIZECLICK;
	else if (s.compare("butregularizeclick") == 0) return BUTREGULARIZECLICK;

	return(0);
};

string PnlPlnrSruRec::VecVDo::getSref(
			const uint ix
		) {
	if (ix == BUTMINIMIZECLICK) return("ButMinimizeClick");
	else if (ix == BUTREGULARIZECLICK) return("ButRegularizeClick");

	return("");
};

/******************************************************************************
 class PnlPlnrSruRec::ContInf
 ******************************************************************************/

PnlPlnrSruRec::ContInf::ContInf(
			const string& TxtRef
		) : Block() {
	this->TxtRef = TxtRef;

	mask = {TXTREF};
};

bool PnlPlnrSruRec::ContInf::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "ContInfPlnrSruRec");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "ContitemInfPlnrSruRec";

	if (basefound) {
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ci", "sref", "TxtRef", TxtRef)) add(TXTREF);
	};

	return basefound;
};

set<uint> PnlPlnrSruRec::ContInf::comm(
			const ContInf* comp
		) {
	set<uint> items;

	if (TxtRef.compare(comp->TxtRef) == 0) insert(items, TXTREF);

	return(items);
};

set<uint> PnlPlnrSruRec::ContInf::diff(
			const ContInf* comp
		) {
	set<uint> commitems;
	set<uint> diffitems;

	commitems = comm(comp);

	diffitems = {TXTREF};
	for (auto it=commitems.begin();it!=commitems.end();it++) diffitems.erase(*it);

	return(diffitems);
};

/******************************************************************************
 class PnlPlnrSruRec::StatApp
 ******************************************************************************/

PnlPlnrSruRec::StatApp::StatApp(
			const bool initdoneDetail
			, const bool initdone1NBlock
			, const bool initdoneSru1NFace
			, const bool initdoneMNLayer
			, const bool initdoneMNStack
		) : Block() {
	this->initdoneDetail = initdoneDetail;
	this->initdone1NBlock = initdone1NBlock;
	this->initdoneSru1NFace = initdoneSru1NFace;
	this->initdoneMNLayer = initdoneMNLayer;
	this->initdoneMNStack = initdoneMNStack;

	mask = {INITDONEDETAIL, INITDONE1NBLOCK, INITDONESRU1NFACE, INITDONEMNLAYER, INITDONEMNSTACK};
};

bool PnlPlnrSruRec::StatApp::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "StatAppPlnrSruRec");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "StatitemAppPlnrSruRec";

	if (basefound) {
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "initdoneDetail", initdoneDetail)) add(INITDONEDETAIL);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "initdone1NBlock", initdone1NBlock)) add(INITDONE1NBLOCK);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "initdoneSru1NFace", initdoneSru1NFace)) add(INITDONESRU1NFACE);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "initdoneMNLayer", initdoneMNLayer)) add(INITDONEMNLAYER);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "initdoneMNStack", initdoneMNStack)) add(INITDONEMNSTACK);
	};

	return basefound;
};

set<uint> PnlPlnrSruRec::StatApp::comm(
			const StatApp* comp
		) {
	set<uint> items;

	if (initdoneDetail == comp->initdoneDetail) insert(items, INITDONEDETAIL);
	if (initdone1NBlock == comp->initdone1NBlock) insert(items, INITDONE1NBLOCK);
	if (initdoneSru1NFace == comp->initdoneSru1NFace) insert(items, INITDONESRU1NFACE);
	if (initdoneMNLayer == comp->initdoneMNLayer) insert(items, INITDONEMNLAYER);
	if (initdoneMNStack == comp->initdoneMNStack) insert(items, INITDONEMNSTACK);

	return(items);
};

set<uint> PnlPlnrSruRec::StatApp::diff(
			const StatApp* comp
		) {
	set<uint> commitems;
	set<uint> diffitems;

	commitems = comm(comp);

	diffitems = {INITDONEDETAIL, INITDONE1NBLOCK, INITDONESRU1NFACE, INITDONEMNLAYER, INITDONEMNSTACK};
	for (auto it=commitems.begin();it!=commitems.end();it++) diffitems.erase(*it);

	return(diffitems);
};

/******************************************************************************
 class PnlPlnrSruRec::StatShr
 ******************************************************************************/

PnlPlnrSruRec::StatShr::StatShr(
			const uint ixPlnrVExpstate
			, const string& scrJrefDetail
			, const string& scrJref1NBlock
			, const bool pnl1nblockAvail
			, const string& scrJrefSru1NFace
			, const string& scrJrefMNLayer
			, const bool pnlmnlayerAvail
			, const string& scrJrefMNStack
			, const bool pnlmnstackAvail
			, const bool ButRegularizeActive
		) : Block() {
	this->ixPlnrVExpstate = ixPlnrVExpstate;
	this->scrJrefDetail = scrJrefDetail;
	this->scrJref1NBlock = scrJref1NBlock;
	this->pnl1nblockAvail = pnl1nblockAvail;
	this->scrJrefSru1NFace = scrJrefSru1NFace;
	this->scrJrefMNLayer = scrJrefMNLayer;
	this->pnlmnlayerAvail = pnlmnlayerAvail;
	this->scrJrefMNStack = scrJrefMNStack;
	this->pnlmnstackAvail = pnlmnstackAvail;
	this->ButRegularizeActive = ButRegularizeActive;

	mask = {IXPLNRVEXPSTATE, SCRJREFDETAIL, SCRJREF1NBLOCK, PNL1NBLOCKAVAIL, SCRJREFSRU1NFACE, SCRJREFMNLAYER, PNLMNLAYERAVAIL, SCRJREFMNSTACK, PNLMNSTACKAVAIL, BUTREGULARIZEACTIVE};
};

bool PnlPlnrSruRec::StatShr::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	string srefIxPlnrVExpstate;

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "StatShrPlnrSruRec");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "StatitemShrPlnrSruRec";

	if (basefound) {
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "srefIxPlnrVExpstate", srefIxPlnrVExpstate)) {
			ixPlnrVExpstate = VecPlnrVExpstate::getIx(srefIxPlnrVExpstate);
			add(IXPLNRVEXPSTATE);
		};
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "scrJrefDetail", scrJrefDetail)) add(SCRJREFDETAIL);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "scrJref1NBlock", scrJref1NBlock)) add(SCRJREF1NBLOCK);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "pnl1nblockAvail", pnl1nblockAvail)) add(PNL1NBLOCKAVAIL);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "scrJrefSru1NFace", scrJrefSru1NFace)) add(SCRJREFSRU1NFACE);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "scrJrefMNLayer", scrJrefMNLayer)) add(SCRJREFMNLAYER);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "pnlmnlayerAvail", pnlmnlayerAvail)) add(PNLMNLAYERAVAIL);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "scrJrefMNStack", scrJrefMNStack)) add(SCRJREFMNSTACK);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "pnlmnstackAvail", pnlmnstackAvail)) add(PNLMNSTACKAVAIL);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "ButRegularizeActive", ButRegularizeActive)) add(BUTREGULARIZEACTIVE);
	};

	return basefound;
};

set<uint> PnlPlnrSruRec::StatShr::comm(
			const StatShr* comp
		) {
	set<uint> items;

	if (ixPlnrVExpstate == comp->ixPlnrVExpstate) insert(items, IXPLNRVEXPSTATE);
	if (scrJrefDetail.compare(comp->scrJrefDetail) == 0) insert(items, SCRJREFDETAIL);
	if (scrJref1NBlock.compare(comp->scrJref1NBlock) == 0) insert(items, SCRJREF1NBLOCK);
	if (pnl1nblockAvail == comp->pnl1nblockAvail) insert(items, PNL1NBLOCKAVAIL);
	if (scrJrefSru1NFace.compare(comp->scrJrefSru1NFace) == 0) insert(items, SCRJREFSRU1NFACE);
	if (scrJrefMNLayer.compare(comp->scrJrefMNLayer) == 0) insert(items, SCRJREFMNLAYER);
	if (pnlmnlayerAvail == comp->pnlmnlayerAvail) insert(items, PNLMNLAYERAVAIL);
	if (scrJrefMNStack.compare(comp->scrJrefMNStack) == 0) insert(items, SCRJREFMNSTACK);
	if (pnlmnstackAvail == comp->pnlmnstackAvail) insert(items, PNLMNSTACKAVAIL);
	if (ButRegularizeActive == comp->ButRegularizeActive) insert(items, BUTREGULARIZEACTIVE);

	return(items);
};

set<uint> PnlPlnrSruRec::StatShr::diff(
			const StatShr* comp
		) {
	set<uint> commitems;
	set<uint> diffitems;

	commitems = comm(comp);

	diffitems = {IXPLNRVEXPSTATE, SCRJREFDETAIL, SCRJREF1NBLOCK, PNL1NBLOCKAVAIL, SCRJREFSRU1NFACE, SCRJREFMNLAYER, PNLMNLAYERAVAIL, SCRJREFMNSTACK, PNLMNSTACKAVAIL, BUTREGULARIZEACTIVE};
	for (auto it=commitems.begin();it!=commitems.end();it++) diffitems.erase(*it);

	return(diffitems);
};

/******************************************************************************
 class PnlPlnrSruRec::Tag
 ******************************************************************************/

PnlPlnrSruRec::Tag::Tag(
			const string& Cpt
		) : Block() {
	this->Cpt = Cpt;

	mask = {CPT};
};

bool PnlPlnrSruRec::Tag::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "TagPlnrSruRec");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "TagitemPlnrSruRec";

	if (basefound) {
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ti", "sref", "Cpt", Cpt)) add(CPT);
	};

	return basefound;
};

/******************************************************************************
 class PnlPlnrSruRec::DpchAppDo
 ******************************************************************************/

PnlPlnrSruRec::DpchAppDo::DpchAppDo(
			const string& scrJref
			, const uint ixVDo
			, const set<uint>& mask
		) : DpchAppPlnr(VecPlnrVDpch::DPCHAPPPLNRSRURECDO, scrJref) {
	if (find(mask, ALL)) this->mask = {SCRJREF, IXVDO};
	else this->mask = mask;

	this->ixVDo = ixVDo;
};

void PnlPlnrSruRec::DpchAppDo::writeXML(
			xmlTextWriter* wr
		) {
	xmlTextWriterStartElement(wr, BAD_CAST "DpchAppPlnrSruRecDo");
	xmlTextWriterWriteAttribute(wr, BAD_CAST "xmlns", BAD_CAST "http://www.epsitechnologies.com/plnr");
		if (has(SCRJREF)) writeString(wr, "scrJref", scrJref);
		if (has(IXVDO)) writeString(wr, "srefIxVDo", VecVDo::getSref(ixVDo));
	xmlTextWriterEndElement(wr);
};

/******************************************************************************
 class PnlPlnrSruRec::DpchEngData
 ******************************************************************************/

PnlPlnrSruRec::DpchEngData::DpchEngData() : DpchEngPlnr(VecPlnrVDpch::DPCHENGPLNRSRURECDATA) {
};

void PnlPlnrSruRec::DpchEngData::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "DpchEngPlnrSruRecData");
	else
		basefound = checkXPath(docctx, basexpath);

	if (basefound) {
		if (extractStringUclc(docctx, basexpath, "scrJref", "", scrJref)) add(SCRJREF);
		if (continf.readXML(docctx, basexpath, true)) add(CONTINF);
		if (statapp.readXML(docctx, basexpath, true)) add(STATAPP);
		if (statshr.readXML(docctx, basexpath, true)) add(STATSHR);
		if (tag.readXML(docctx, basexpath, true)) add(TAG);
	} else {
		continf = ContInf();
		statapp = StatApp();
		statshr = StatShr();
		tag = Tag();
	};
};

File: webappplnr/CrdPlnrSru/PnlPlnrSruRec.js
/**
  * \file PnlPlnrSruRec.js
  * web client functionality for panel PnlPlnrSruRec
  * \author Alexander Wirthmueller
  * \date created: 4 Dec 2015
  * \date modified: 4 Dec 2015
  */

function updateScrJrefs() {
	scrJrefDetail = retrieveSi(srcdoc, "StatShrPlnrSruRec", "scrJrefDetail");
	scrJref1NBlock = retrieveSi(srcdoc, "StatShrPlnrSruRec", "scrJref1NBlock");
	scrJrefSru1NFace = retrieveSi(srcdoc, "StatShrPlnrSruRec", "scrJrefSru1NFace");
	scrJrefMNLayer = retrieveSi(srcdoc, "StatShrPlnrSruRec", "scrJrefMNLayer");
	scrJrefMNStack = retrieveSi(srcdoc, "StatShrPlnrSruRec", "scrJrefMNStack");
};

function resetInitdones() {
	setSi(srcdoc, "StatAppPlnrSruRec", "initdoneDetail", "false");
	setSi(srcdoc, "StatAppPlnrSruRec", "initdone1NBlock", "false");
	setSi(srcdoc, "StatAppPlnrSruRec", "initdoneSru1NFace", "false");
	setSi(srcdoc, "StatAppPlnrSruRec", "initdoneMNLayer", "false");
	setSi(srcdoc, "StatAppPlnrSruRec", "initdoneMNStack", "false");
};

function resetHeights() {
	heightDetail = 30;
	height1NBlock = 30;
	heightSru1NFace = 30;
	heightMNLayer = 30;
	heightMNStack = 30;
};

function getInitdone(pnlshort) {
	return(retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdone" + pnlshort) == "true");
};

function setInitdone(pnlshort) {
	setSi(srcdoc, "StatAppPlnrSruRec", "initdone" + pnlshort, "true");

	checkInitdone();
};

function checkInitdone() {
	var initdoneDetail = (retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdoneDetail") == "true");
	var initdoneAPar = (retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdoneAPar") == "true");
	var initdone1NRelease = (retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdone1NRelease") == "true");

	var initdoneDetail = (retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdoneDetail") == "true");
	var initdone1NBlock = (retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdone1NBlock") == "true");
	var initdoneSru1NFace = (retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdoneSru1NFace") == "true");
	var initdoneMNLayer = (retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdoneMNLayer") == "true");
	var initdoneMNStack = (retrieveSi(srcdoc, "StatAppPlnrSruRec", "initdoneMNStack") == "true");

	if (!initdoneDetail) {
		lhsdoc.getElementById("Detail").src = "./PnlPlnrSruDetail.html?scrJref=" + scrJrefDetail;
	} else if (!initdone1NBlock) {
		rhsdoc.getElementById("1NBlock").src = "./PnlPlnrSru1NBlock.html?scrJref=" + scrJref1NBlock;
	} else if (!initdoneSru1NFace) {
		rhsdoc.getElementById("Sru1NFace").src = "./PnlPlnrSruSru1NFace.html?scrJref=" + scrJrefSru1NFace;
	} else if (!initdoneMNLayer) {
		rhsdoc.getElementById("MNLayer").src = "./PnlPlnrSruMNLayer.html?scrJref=" + scrJrefMNLayer;
	} else if (!initdoneMNStack) {
		rhsdoc.getElementById("MNStack").src = "./PnlPlnrSruMNStack.html?scrJref=" + scrJrefMNStack;

	} else {
		refreshB();
	};
};

function setPnlAvail(short, avail) {
	var lhsrhsdoc;

	if ((short == "Detail")) lhsrhsdoc = lhsdoc;
	else lhsrhsdoc = rhsdoc;

	var oldAvail = (lhsrhsdoc.getElementById("tr" + short).getAttribute("class") == "show");

	var height;

	if (avail != oldAvail) {
		if (avail) {
			height = 30;

			lhsrhsdoc.getElementById("spc" + short).setAttribute("class", "show");
			lhsrhsdoc.getElementById("tr" + short).setAttribute("class", "show");

		} else {
			height = -13;

			lhsrhsdoc.getElementById(short).contentWindow.minimize(false);
			
			lhsrhsdoc.getElementById("spc" + short).setAttribute("class", "hide");
			lhsrhsdoc.getElementById("tr" + short).setAttribute("class", "hide");
		};

		if (short == "Headbar") heightHeadbar = height;
		else if (short == "List") heightList = height;
		else if (short == "Rec") heightRec = height;
		else if (short == "Detail") heightDetail = height;
		else if (short == "1NBlock") height1NBlock = height;
		else if (short == "Sru1NFace") heightSru1NFace = height;
		else if (short == "MNLayer") heightMNLayer = height;
		else if (short == "MNStack") heightMNStack = height;
	};

	return(avail != oldAvail);
};

// --- expand state management
function minimize() {
	resetInitdones();

	getCrdwnd().changeHeight("Rec", 45);
	doc.getElementById("tdSide").setAttribute("height", "45");
	doc.getElementById("Rec_side").setAttribute("height", "45");
	doc.getElementById("Rec_cont").setAttribute("height", "45");

	doc.getElementById("Rec_side").src = "./PnlPlnrSruRec_aside.html";
	doc.getElementById("Rec_cont").src = "./PnlPlnrSruRec_a.html";
};

function regularize() {
	resetHeights();

	getCrdwnd().changeHeight("Rec", 223);
	doc.getElementById("tdSide").setAttribute("height", "223");
	doc.getElementById("Rec_side").setAttribute("height", "223");
	doc.getElementById("Rec_cont").setAttribute("height", "223");

	doc.getElementById("Rec_side").src = "./PnlPlnrSruRec_bside.html";
	doc.getElementById("Rec_cont").src = "./PnlPlnrSruRec_b.html";
};

function changeHeight(pnlshort, height, update) {
	var lhsrhsdoc;

	if ((pnlshort == "Detail")) lhsrhsdoc = lhsdoc;
	else lhsrhsdoc = rhsdoc;

	lhsrhsdoc.getElementById("td" + pnlshort).setAttribute("height", "" + height);
	lhsrhsdoc.getElementById(pnlshort).setAttribute("height", "" + height);

	if (pnlshort == "Headbar") heightHeadbar = height;
	else if (pnlshort == "List") heightList = height;
	else if (pnlshort == "Rec") heightRec = height;
	else if (pnlshort == "Detail") heightDetail = height;
	else if (pnlshort == "1NBlock") height1NBlock = height;
	else if (pnlshort == "Sru1NFace") heightSru1NFace = height;
	else if (pnlshort == "MNLayer") heightMNLayer = height;
	else if (pnlshort == "MNStack") heightMNStack = height;

	if (update) updateHeight();
};

function updateHeight() {
	var heightLhs, heightRhs, heightGt;

	heightLhs = heightDetail+13 + 5;
	heightRhs = height1NBlock+13 + heightSru1NFace+13 + heightMNLayer+13 + heightMNStack+13 + 5;

	if (heightLhs > heightRhs) {
		lhsdoc.getElementById("tdFill").setAttribute("height", "5");
		rhsdoc.getElementById("tdFill").setAttribute("height", "" + (heightLhs-heightRhs+5));

		heightGt = heightLhs;

	} else {
		lhsdoc.getElementById("tdFill").setAttribute("height", "" + (heightRhs-heightLhs+5));
		rhsdoc.getElementById("tdFill").setAttribute("height", "5");

		heightGt = heightRhs;
	};

	getCrdwnd().changeHeight("Rec", 45+heightGt+1);

	doc.getElementById("tdSide").setAttribute("height", "" + (45+heightGt+1));
	doc.getElementById("Rec_side").setAttribute("height", "" + (45+heightGt+1));
	doc.getElementById("Rec_cont").setAttribute("height", "" + (45+heightGt+1));

	sidedoc.getElementById("tdFlex").setAttribute("height", "" + (heightGt+1));

	contdoc.getElementById("tdLhs").setAttribute("height", "" + heightGt);
	contdoc.getElementById("Rec_lhs").setAttribute("height", "" + heightGt);
	contdoc.getElementById("tdRhs").setAttribute("height", "" + heightGt);
	contdoc.getElementById("Rec_rhs").setAttribute("height", "" + heightGt);
};

// --- view initialization and refresh
function initAside() {
	if (!doc) return;

	sidedoc = doc.getElementById("Rec_side").contentDocument;

	refreshAside();
};

function initA() {
	if (!doc) return;

	contdoc = doc.getElementById("Rec_cont").contentDocument;

	setTextContent(contdoc, contdoc.getElementById("Cpt"), retrieveTi(srcdoc, "TagPlnrSruRec", "Cpt"));

	refreshA();
};

function initB() {
	if (!doc) return;

	sidedoc = doc.getElementById("Rec_side").contentDocument;
	contdoc = doc.getElementById("Rec_cont").contentDocument;
	hdrdoc = contdoc.getElementById("Rec_hdr").contentDocument;
	lhsdoc = contdoc.getElementById("Rec_lhs").contentDocument;
	rhsdoc = contdoc.getElementById("Rec_rhs").contentDocument;

	setTextContent(hdrdoc, hdrdoc.getElementById("Cpt"), retrieveTi(srcdoc, "TagPlnrSruRec", "Cpt"));

	checkInitdone();
};

function init() {
	var srefIxPlnrVExpstate = retrieveSi(srcdoc, "StatShrPlnrSruRec", "srefIxPlnrVExpstate");

	if (srefIxPlnrVExpstate == "mind") {
		minimize();
	} else if (srefIxPlnrVExpstate == "regd") {
		regularize();
	};
};

function refreshAside() {
	var ButRegularizeActive;

	ButRegularizeActive = (retrieveSi(srcdoc, "StatShrPlnrSruRec", "ButRegularizeActive") == "true");

	if (ButRegularizeActive) {
		sidedoc.getElementById("ButRegularize").src = "../icon/regularize.png";
		sidedoc.getElementById("ButRegularize").setAttribute("class", "side_pm");
		sidedoc.getElementById("ButRegularize").setAttribute("onmouseover", "src='../icon/regularize_hlt.png'");
		sidedoc.getElementById("ButRegularize").setAttribute("onmouseout", "src='../icon/regularize.png'");
		sidedoc.getElementById("ButRegularize").setAttribute("onclick", "handleButRegularizeClick()");
	} else {
		sidedoc.getElementById("ButRegularize").src = "../icon/regularize_inact.png";
		sidedoc.getElementById("ButRegularize").setAttribute("class", "side_pmdis");
		sidedoc.getElementById("ButRegularize").setAttribute("onmouseover", "");
		sidedoc.getElementById("ButRegularize").setAttribute("onmouseout", "");
		sidedoc.getElementById("ButRegularize").setAttribute("onclick", "");
	};
};

function refreshA() {
	setTextContent(contdoc, contdoc.getElementById("TxtRef"), retrieveCi(srcdoc, "ContInfPlnrSruRec", "TxtRef"));
};

function refreshB() {
	var updh = false;

	var pnl1nblockAvail = (retrieveSi(srcdoc, "StatShrPlnrSruRec", "pnl1nblockAvail") == "true");
	var pnlmnlayerAvail = (retrieveSi(srcdoc, "StatShrPlnrSruRec", "pnlmnlayerAvail") == "true");
	var pnlmnstackAvail = (retrieveSi(srcdoc, "StatShrPlnrSruRec", "pnlmnstackAvail") == "true");

	setTextContent(hdrdoc, hdrdoc.getElementById("TxtRef"), retrieveCi(srcdoc, "ContInfPlnrSruRec", "TxtRef"));

	if (setPnlAvail("1NBlock", pnl1nblockAvail)) updh = true;
	if (setPnlAvail("MNLayer", pnlmnlayerAvail)) updh = true;
	if (setPnlAvail("MNStack", pnlmnstackAvail)) updh = true;

	if (updh) updateHeight();
};

function refresh() {
	var srefIxPlnrVExpstate = retrieveSi(srcdoc, "StatShrPlnrSruRec", "srefIxPlnrVExpstate");

	if (srefIxPlnrVExpstate == "mind") {
		refreshAside();
		refreshA();
	} else if (srefIxPlnrVExpstate == "regd") {
		refreshB();
	};
};

// --- event handlers
function handleLoad() {
	scrJref = location.search.substr(location.search.indexOf("scrJref=")+8);

	doc = document;
	srcdoc = doc.getElementById("Rec_src").contentDocument;

	var str = serializeDpchApp("DpchAppPlnrInit", scrJref);
	sendReq(str, doc, handleDpchAppInitReply, true);
};

// --- specific event handlers for app controls

// --- generalized event handlers for app controls

// --- generalized event handlers for shared controls

function handleButClick(ctlsref) {
	var str = serializeDpchAppDo(srcdoc, "DpchAppPlnrSruRecDo", scrJref, ctlsref + "Click");
	sendReq(str, doc, handleDpchAppDataDoReply);
};

// --- server interaction
function mergeDpchEngData(dom) {
	var mask = [];

	if (updateSrcblock(dom, "DpchEngPlnrSruRecData", "ContInfPlnrSruRec", srcdoc)) mask.push("continf");
	if (updateSrcblock(dom, "DpchEngPlnrSruRecData", "StatAppPlnrSruRec", srcdoc)) mask.push("statapp");
	if (updateSrcblock(dom, "DpchEngPlnrSruRecData", "StatShrPlnrSruRec", srcdoc)) mask.push("statshr");
	if (updateSrcblock(dom, "DpchEngPlnrSruRecData", "TagPlnrSruRec", srcdoc)) mask.push("tag");

	return mask;
};

function handleDpchEng(dom, dpch) {
	var _scrJref, mask;
	var srefIxPlnrVExpstate, newSrefIxPlnrVExpstate;

	_scrJref = retrieveValue(dom, "//plnr:*/plnr:scrJref");
	srefIxPlnrVExpstate = retrieveSi(srcdoc, "StatShrPlnrSruRec", "srefIxPlnrVExpstate");

	if (_scrJref == scrJref) {
		if (dpch == "DpchEngPlnrSruRecData") {
			mask = mergeDpchEngData(dom);

			if (mask.indexOf("statshr") != -1) {
				newSrefIxPlnrVExpstate = retrieveSi(srcdoc, "StatShrPlnrSruRec", "srefIxPlnrVExpstate");

				if (newSrefIxPlnrVExpstate != srefIxPlnrVExpstate) {
					updateScrJrefs();
					
					if (newSrefIxPlnrVExpstate == "mind") {
						minimize();
					} else if (newSrefIxPlnrVExpstate == "regd") {
						regularize();
					};

				} else {
					refresh();
				};

			} else {
				refresh();
			};
		};

	} else {
		if (srefIxPlnrVExpstate != "mind") {
			if (_scrJref == scrJrefDetail) {
				lhsdoc.getElementById("Detail").contentWindow.handleDpchEng(dom, dpch);
			} else if (_scrJref == scrJref1NBlock) {
				rhsdoc.getElementById("1NBlock").contentWindow.handleDpchEng(dom, dpch);
			} else if (_scrJref == scrJrefSru1NFace) {
				rhsdoc.getElementById("Sru1NFace").contentWindow.handleDpchEng(dom, dpch);
			} else if (_scrJref == scrJrefMNLayer) {
				rhsdoc.getElementById("MNLayer").contentWindow.handleDpchEng(dom, dpch);
			} else if (_scrJref == scrJrefMNStack) {
				rhsdoc.getElementById("MNStack").contentWindow.handleDpchEng(dom, dpch);
			} else {
				// alert("got a '" + dpch + "' from job with scrJref " + _scrJref);
			};

		} else {
			// alert("got a '" + dpch + "' from job with scrJref " + _scrJref);
		};
	};
};

function handleDpchAppInitReply() {
	var dom, blk;

	if (doc.req.readyState == 4) {
		dom = doc.req.responseXML;

		blk = retrieveBlock(dom, "//plnr:*");
		if (blk) {
			if (blk.nodeName == "DpchEngPlnrSruRecData") {
				mergeDpchEngData(dom);

				updateScrJrefs();
				init();

				getCrdwnd().setInitdone("Rec");
			};
		};
	};
};

function handleDpchAppDataDoReply() {
	var dom, blk, mask;
	var srefIxPlnrVExpstate, newSrefIxPlnrVExpstate;

	if (doc.req.readyState == 4) {
		dom = doc.req.responseXML;

		// check dispatch type
		blk = retrieveBlock(dom, "//plnr:*");

		if (blk) {
			if (blk.nodeName == "DpchEngPlnrConfirm") {
				// do nothing

			} else if (blk.nodeName == "DpchEngPlnrSruRecData") {
				srefIxPlnrVExpstate = retrieveSi(srcdoc, "StatShrPlnrSruRec", "srefIxPlnrVExpstate");

				mask = mergeDpchEngData(dom);

				if (mask.indexOf("statshr") != -1) {
					newSrefIxPlnrVExpstate = retrieveSi(srcdoc, "StatShrPlnrSruRec", "srefIxPlnrVExpstate");

					if (newSrefIxPlnrVExpstate != srefIxPlnrVExpstate) {
						updateScrJrefs();

						if (newSrefIxPlnrVExpstate == "mind") {
							minimize();
						} else if (newSrefIxPlnrVExpstate == "regd") {
							regularize();
						};
					};

				} else {
					refresh();
				};
			};
		};
	};
};


